store/watcher.go

```diff
@@ -2,59 +2,170 @@ package store
+type Registration interface {
+	EventChan() <-chan *clientv3.Event
+	Unregister()
+}
+
+type registration struct {
+	eventChan chan *clientv3.Event
+	Watcher   Watcher
+	key       string
+}
+
+func (r registration) EventChan() <-chan *clientv3.Event {
+	return r.eventChan
+}
+
+func (r registration) Unregister() {
+	r.Watcher.unregister(r.key)
+}
+
+type Watcher struct {
+	config         *config.Config
+	prefix         string
+	etcdWatcher    EtcdWatcher
+	registrations  map[string]registration
+	registrationsM *sync.RWMutex
+}
+
+type WatcherOpt func(w *Watcher)
+
+func WithPrefix(prefix string) WatcherOpt {
+	return func(w *Watcher) {
+		prefix = prefixedKey(w.config, prefix)
+		w.prefix = prefix
+	}
+}
+
+func WithEtcdWatcher(etcdWatcher EtcdWatcher) WatcherOpt {
+	return func(w *Watcher) {
+		w.etcdWatcher = etcdWatcher
+	}
+}
+
+func NewWatcher(ctx context.Context, config *config.Config, opts ...WatcherOpt) (Watcher, error) {
 	log := logger.Get(ctx)
-	prefix = s.Key(prefix)
+	w := Watcher{
+		config:         config,
+		registrations:  make(map[string]registration),
+		registrationsM: &sync.RWMutex{},
+	}
 
-	client, err := etcd.NewClient()
-	if err != nil {
-		return nil, errors.Wrapf(err, "fail to create etcd client")
+	for _, opt := range opts {
+		opt(&w)
 	}
 
-	wc := clientv3.NewWatcher(client)
-	log.Infof("create endpoints watcher on prefix %v", prefix)
+	if w.prefix == "" {
+		w.prefix = prefixedKey(w.config, "/")
+	}
+
+	log.Infof("create endpoints Watcher on prefix %v", w.prefix)
 
-	// Use context.Background() to avoid the resulting chan to be closed at the end of a HTTP request
-	// https://godoc.org/go.etcd.io/etcd/clientv3#Watcher
-	wchan := wc.Watch(context.Background(), prefix, clientv3.WithPrefix())
+	if w.etcdWatcher == nil {
+		etcdWatcher, err := etcd.NewWatcher(w.prefix)
+		if err != nil {
+			return Watcher{}, errors.Wrapf(err, "fail to create etcd Watcher on %v", w.prefix)
+		}
+		w.etcdWatcher = etcdWatcher
+	}
 
-	return watcher{etcdClient: client, etcdWatcher: wc, etcdWatchChan: wchan, ready: make(chan struct{})}, nil
+	go func() {
+		w.watchModifications(ctx)
+	}()
+
+	return w, nil
 }
 
-type watcher struct {
-	etcdClient    *clientv3.Client
-	etcdWatcher   clientv3.Watcher
-	etcdWatchChan clientv3.WatchChan
-	ready         chan struct{}
+func (w Watcher) watchModifications(ctx context.Context) {
+	log := logger.Get(ctx)
+	for res := range w.etcdWatcher.WatchChan() {
+		if err := res.Err(); err != nil {
+			// If the connection is canceled because grpc (HTTP/2) connection is
+			// closed as etcd restart We don't want to throw an error but just keep
+			// looping as the client will automatically reconnect.
+			if etcderr, ok := err.(rpctypes.EtcdError); ok && etcderr.Code() == codes.Canceled {
+				log.WithError(err).Info("watch response canceled, retry")
+			} else if err != nil {
+				log.WithError(err).Error("fail to handle Watcher response")
```

Comment:
Don't we want to `continue` in this case?

Reply:
You're absolutely right, thanks a lot, I'm going to add a spec